core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -177,13 +227,30 @@ object ByteBufferMessageSet {
     buffer.putInt(messageWriter.size)
     messageWriter.writeTo(buffer)
   }
+
+
+  case class FilterResult(messagesRead: Int,
+                          bytesRead: Int,
+                          messagesRetained: Int,
+                          bytesRetained: Int,
+                          maxTimestamp: Long,
+                          offsetOfMaxTimestamp: Long)
 }
 
 private object OffsetAssigner {
 
   def apply(offsetCounter: LongRef, size: Int): OffsetAssigner =
     new OffsetAssigner(offsetCounter.value to offsetCounter.addAndGet(size))
 
+  def apply(baseOffset: Long, magic: Byte, messageAndOffsets: Seq[MessageAndOffset]): OffsetAssigner =
+    new OffsetAssigner(messageAndOffsets.map { messageAndOffset =>
+      if (magic > Message.MagicValue_V0)
+        // The offset of the messages are absolute offset, compute the inner offset.
+        messageAndOffset.offset - baseOffset
```

Comment:
Hmm, will `OffsetAssigner.nextAbsoluteOffset` return an absolute offset if we do this?